lxd/device/cdi/configure.go

```diff
@@ -0,0 +1,335 @@
+		tegraCSVFiles := make([]string, 0)
+		for _, candidate := range tegraCSVFilesCandidates {
+			_, err := os.Stat(candidate)
+			if err == nil {
+				tegraCSVFiles = append(tegraCSVFiles, candidate)
+			} else if os.IsNotExist(err) {
+				continue
+			} else {
+				return nil, err
+			}
+		}
+
+		if len(tegraCSVFiles) == 0 {
+			return nil, fmt.Errorf("No CSV files detected for Tegra iGPU")
+		}
+
+		for _, tegraFile := range tegraCSVFiles {
+			configDevices.BindMounts = append(
+				configDevices.BindMounts,
+				map[string]string{
+					"type":     "disk",
+					"source":   tegraFile,
+					"path":     strings.TrimPrefix(tegraFile, rootPath),
+					"shift":    shift,
+					"readonly": "true",
+				},
+			)
+		}
+	}
+
+	return indirectSymlinks, nil
+}
+
+// specHookToLXDCDIHook will translate a hook from a CDI spec into an entry in a `Hooks`.
+// Some CDI hooks are not relevant for LXD and will be ignored.
+func specHookToLXDCDIHook(hook *specs.Hook, hooks *Hooks, l logger.Logger) error {
+	if hook == nil {
+		l.Warn("CDI hook is nil")
+		return nil
+	}
+
+	rootPath := ""
+	if shared.InSnap() {
+		rootPath = "/var/lib/snapd/hostfs"
+	}
+
+	if len(hook.Args) < 3 {
+		return fmt.Errorf("Not enough arguments for CDI hook: %v", hook.Args)
+	}
+
+	processCreateSymlinksHook := func(args []string) error {
+		if len(args) < 2 {
+			return fmt.Errorf("CDI 'create-symlinks' hook has not enough arguments: %v", hook.Args)
+		}
+
+		// The list of arguments is either
+		// `--link <target>::<link> --link <target>::<link> ...`
+		// or `--link=<target>::<link> --link=<target>::<link> ...`
+		// and we need to handle both cases as they are both valid.
+		var targetWithLink string
+		for i := 0; i < len(args); i += 1 {
+			if args[i] == "--link" {
+				continue
+			}
+
+			if strings.Contains(args[i], "=") {
+				// We can assume the arg is `--link=<target>::<link>`
+				splitted := strings.Split(args[i], "=")
+				if len(splitted) != 2 {
+					return fmt.Errorf("Invalid symlink arg %q", args[i])
+				}
+
+				targetWithLink = splitted[1]
+			} else {
+				// We can assume the arg is `<target>::<link>`
+				targetWithLink = args[i]
+			}
+
+			entry := strings.Split(targetWithLink, "::")
+			if len(entry) != 2 {
+				return fmt.Errorf("Invalid symlink entry %q", targetWithLink)
+			}
+
+			// `Link` is always an absolute path and `Target` (a `Link` points to a `Target`) is relative
+			// to the `Link` location in the CDI spec. A resolving operation will be needed to have the absolute
+			// path of the `Target`
+			hooks.Symlinks = append(hooks.Symlinks, SymlinkEntry{Target: strings.TrimPrefix(entry[0], rootPath), Link: strings.TrimPrefix(entry[1], rootPath)})
+		}
+
+		return nil
+	}
+
+	processUpdateLdcacheHook := func(args []string) error {
+		// The `update-ldcache` hook can be used without any arguments, but in our case we don't need it.
+		if len(args) < 2 {
+			logger.Warnf("CDI 'update-ldcache' hook has not enough arguments: %v", args)
```

Comment:
why not using the logger passed in here? 

Also we dont use `logger.*f()` functions, please use contextual logger.

Reply:
Why is this scenario not an error?

Reply:
Actually this is a mistake from me: the case `nvidia-ctk hook update-ldcache --folder=<entry>` (in this case `len(args) == 1`) is valid, but here it'll be ignored.. Fixing that (I'll simply remove this if block and it will be good). Thanks!

Reply:
fixed